File: 051 N-Queens/n_queens.java
public class Solution {
    private char[] template;
    private int rows;
    private List<List<String>> result;
    public List<List<String>> solveNQueens(int n) {
        template = new char[n];
        rows = n;
        result = new ArrayList<List<String>>();
        for (int i = 0; i < n; i++) template[i] = '.';
        int[] row = new int[n];
        solve(row, 0);
        return result;
    }
    
    private void solve(int[] row, int curr) {
        if (curr == rows) {
            result.add(generate(row));
            return;
        }
        for (int i = 0; i < rows; i++) {
            row[curr] = i;
            if (check(row, curr, i))
                solve(row, curr + 1);
            row[curr] = 0;
        }
    }
    
    private boolean check(int[] row, int r, int c) {
        for (int i = 0; i < r; i++) {
            if (row[i] == c) return false;
            if (Math.abs(row[i] - c) == Math.abs(i - r)) return false;
        }
        return true;
    }
    
    private List<String> generate(int[] row) {
        List<String> res = new ArrayList<String>();
        for (int i = 0; i < rows; i++) {
            char[] line = template.clone();
            line[row[i]] = 'Q';
            String s = new String(line);
            res.add(s);
        }
        return res;
    }
} // Author: XC

File: 051 N-Queens/n_queens.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
		rows = cols = n;
		vector<string> board(rows, string(cols, '.'));
		dfs(0, board);
		return result;
	}

	void dfs(int row, vector<string>& board) {
		if (row == rows) {
			result.push_back(board);
			return;
		}
		for (int col = 0; col < cols; ++col) {
			if (check_board(board, row, col)) {
				board[row][col] = 'Q';
				dfs(row + 1, board);
				board[row][col] = '.';
			}
		}
	}
	
	bool check_board(vector<string>& board, int row, int col) {
		for (int i = 0; i != row; ++i) {
			if (board[i][col] == 'Q') return false;
		}
		for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j) {
			if (board[i][j] == 'Q') return false;
		}
		for (int i = row - 1, j = col + 1; i >= 0 && j < cols; --i, ++j) {
			if (board[i][j] == 'Q') return false;
		}
		return true;
	}

	void print_board(vector<string>& board) {
		for (auto row : board) {
			cout << row << endl;
		}
		cout << endl;
	}
private:
	vector<vector<string>> result;
	int rows, cols;
}; // Author: XC